File: codeforces/652/D/D.cpp
#include <bits/stdc++.h>

using namespace std;

constexpr size_t MAX_N = 2000001;
constexpr long long MODULUS = (1e9 + 7);

int memoT[MAX_N];
int memoF[MAX_N];

int ans(int i, bool disallow_occupy) {
  if (i <= 2) return 0;
  if (i == 3) {
    if (disallow_occupy) return 0;
    return 4;
  }
  if (i == 4) return 4;

  if (disallow_occupy && memoT[i] != -1) return memoT[i];
  if (!disallow_occupy && memoF[i] != -1) return memoF[i];

  int choice1 = (2LL*ans(i-2, true) + ans(i-1, true) + 4) % MODULUS;

  if (disallow_occupy) choice1 = -999;

  int choice2 = (2LL*ans(i-2, false) + ans(i-1, false)) % MODULUS;
  choice2 = choice2 % MODULUS;

  if (disallow_occupy) {
    return memoT[i] = choice2;
  } else {
    return memoF[i] = max(choice1, choice2);
  }

}

int main() {
  
  for (int i=0; i<MAX_N; i++) {
    memoT[i] = -1;
    memoF[i] = -1;
  }
  for (int i=5; i<MAX_N; i++) {
    ans(i, true);
    ans(i, false);
  }

  int T;

  cin >> T;

  while (T --> 0) {
    int n;
    cin >> n;
    cout << " " << max(ans(n, true), ans(n, false)) << endl;    
  }  
  return 0;
}

File: codeforces/652/D/D.py
#!/usr/bin/env pypy3

import array

MODULUS = 10**9+7

from sys import stdin, stdout
 
def input():
    return stdin.readline().strip()

memoT = array.array('i', [-1]*(2*10**6+1))
memoF = array.array('i', [-1]*(2*10**6+1))

def ans(i, disallow_occupy):
    if i <= 2: return 0
    if i == 3: 
        if disallow_occupy: return 0
        return 4
    if i == 4:
        return 4

    if disallow_occupy and memoT[i] != -1: return memoT[i]
    if not disallow_occupy and memoF[i] != -1: return memoF[i]

    choice1 = 2*ans(i-2, True) + ans(i-1, True) + 4 
    choice1 = choice1 % MODULUS
    if disallow_occupy: choice1 = float("-inf")

    choice2 = 2*ans(i-2, False) + ans(i-1, False)
    choice2 = choice2 % MODULUS

    if disallow_occupy:
        memoT[i] = max(choice1, choice2)
        return memoT[i]
    else:
        memoF[i] = max(choice1, choice2)
        return memoF[i]    

for i in range(5, 100):
    print(f"{i} {ans(i, True)} {ans(i, False)}")
